File: examples/client/python/contact-list.py
#!/usr/bin/env python

import os
from gi.repository import GObject
GObject.threads_init()

from gi.repository import TelepathyGLib as Tp

def manager_prepared_cb(manager, result, loop):
    manager.prepare_finish(result)

    for account in manager.get_valid_accounts():
        connection = account.get_connection()

        # Verify account is online and received its contact list. If state is not
        # SUCCESS this means we didn't received the roster from server yet and
        # we would have to wait for the "notify:contact-list-state" signal. */
        if connection is not None and \
           connection.get_contact_list_state() == Tp.ContactListState.SUCCESS:
            contacts = connection.dup_contact_list()
            for contact in contacts:
                print "%s (%s)" % (contact.get_identifier(), contact.get_contact_groups())
    loop.quit()

if __name__ == '__main__':
    Tp.debug_set_flags(os.getenv('EXAMPLE_DEBUG', ''))

    loop = GObject.MainLoop()
    manager = Tp.AccountManager.dup()
    factory = manager.get_factory()
    factory.add_account_features([Tp.Account.get_feature_quark_connection()])
    factory.add_connection_features([Tp.Connection.get_feature_quark_contact_list()])
    factory.add_contact_features([Tp.ContactFeature.CONTACT_GROUPS])

    manager.prepare_async(None, manager_prepared_cb, loop)
    loop.run()

File: examples/client/js/contact-list.js
#! /usr/bin/gjs

const Tp = imports.gi.TelepathyGLib;
const Mainloop = imports.mainloop;

let manager = Tp.AccountManager.dup();
let factory = manager.get_factory();
factory.add_account_features([Tp.Account.get_feature_quark_connection()]);
factory.add_connection_features([Tp.Connection.get_feature_quark_contact_list()]);
factory.add_contact_features([Tp.ContactFeature.CONTACT_GROUPS]);

manager.prepare_async(null, function(self, result) {
    manager.prepare_finish(result);

    let accounts = manager.get_valid_accounts();
    for (let i = 0; i < accounts.length; i++) {
        let connection = accounts[i].get_connection();

        if (connection != null &&
            connection.get_contact_list_state() == Tp.ContactListState.SUCCESS) {

            let contacts = connection.dup_contact_list();
            for (let j = 0; j < contacts.length; j++) {
                let contact = contacts[j];
                // poor man's printf...
                print (contact.get_identifier() + ' (' + contact.get_contact_groups() + ')');
            }
        }
    }
    Mainloop.quit('example');
});

Mainloop.run('example');
